onnxruntime/core/providers/webnn/builders/impl/expand_op_builder.cc

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) Intel Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/safeint.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/optimizer/initializer.h"
+#include "core/providers/common.h"
+#include "core/providers/cpu/tensor/reshape_helper.h"
+#include "core/providers/shared/utils/utils.h"
+#include "core/providers/webnn/builders/helper.h"
+#include "core/providers/webnn/builders/model_builder.h"
+#include "core/providers/webnn/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace webnn {
+
+class ExpandOpBuilder : public BaseOpBuilder {
+  // Add operator related.
+ public:
+  void AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const override;
+
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+
+  // Operator support related.
+ private:
+  bool IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                         const logging::Logger& logger) const override;
+};
+
+bool GetExpandShape(const onnx::TensorProto& tensor, std::vector<int64_t>& shape, const logging::Logger& logger) {
+  std::vector<uint8_t> unpacked_tensor;
+  auto status = onnxruntime::utils::UnpackInitializerData(tensor, unpacked_tensor);
+  if (!status.IsOK()) {
+    LOGS(logger, ERROR) << "Error while unpacking shape: " << status.ErrorMessage();
+    return false;
+  }
+  const auto& dims = tensor.dims();
+  if (dims.empty() || dims[0] == 0) {
+    LOGS(logger, VERBOSE) << "The shape of expand cannot be empty.";
+    return false;
+  }
+  if (dims.size() != 1) {
+    LOGS(logger, VERBOSE) << "The shape of expand must be 1D.";
+    return false;
+  }
+  const int64_t* shape_data = reinterpret_cast<const int64_t*>(unpacked_tensor.data());
+  shape = std::vector<int64_t>{shape_data, shape_data + dims[0]};
+  return true;
+}
+
+// Add operator related.
+
+void ExpandOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  model_builder.AddInitializerToSkip(node.InputDefs()[1]->Name());
+}
+
+Status ExpandOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                              const Node& node,
+                                              const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  const auto& initializers(model_builder.GetInitializerTensors());
+  const auto& shape_tensor = *initializers.at(input_defs[1]->Name());
+  std::vector<int64_t> raw_shape;
+  // We already checked the raw_shape (new_shape) in IsOpSupportedImpl.
+  GetExpandShape(shape_tensor, raw_shape, logger);
```

Comment:
`GetExpandShape` return value not checked.

Reply:
Can we enlarge `raw_shape` to input.rank, right aligned with leading ones before passing to WebNN? My implementation of expand requires that the input rank and new shape length match (like slice).

Reply:
> Can we enlarge raw_shape to input.rank, right aligned with leading ones before passing to WebNN? 

On the above case, the size of new shape should be less than the size of input shape. So the size of new shape being greater than the size of input shape isn't allowed currently, right?

Reply:
> So the size of new shape being greater than the size of input shape isn't allowed currently, right?

Yes, the old input shape and new output shape must have the same dimension count (or rank), where `raw_shape.size() == input.shape().size() == output.shape.size()`. Did you encounter a case in the model where the input shape and new output shape had differing rank? (the cases I've seen match, so if the input is 3D, then the new shape has 3 elements)

Reply:
Is there a case of the size of new shape being greater than the one of input's shape?
For example 2D to 3D,
```
input_shape = [a, b];
new_shape = [1, a, b];
```

Reply:
> Is there a case of the size of new shape being greater than the one of input's shape?

Although I haven't seen it in real models, ONNX allows it (https://github.com/onnx/onnx/blob/main/docs/Operators.md#expand in the dim_changed example), and so probably there exists at least one model out there which does. I would *prefer* to see the WebNN implementation be simple in its rules (like slice which requires `starts` and `sizes` to match rank with `input`) and for frameworks to resolve these higher layer policies before reaching the lower level WebNN. Is it easy to call `reshape` on the input with leading 1's if the `newShape` has more dimensions?